### PR TITLE
Plugin class

### PR DESCRIPTION
**Pull request by sofroniewn**
# Description

Based on conversations with @royerloic and @kne42 this is an incredibly early and raw attempt at making a plugin class that could be added to the viewer. It is meant to enable some initial investigations into this idea rather than a long term solution. Right now there is a very simple plugin base class that all plugins can inherit from and a plugin Qt class that plugin Qt elements can inherit from. The plugin class has this Qt class on it as an `_qt` element and when the plugin gets added to the viewer the `_qt` element will populate the viewer in the appropriate place.

A long term goal is to support creating plugins out of function annotations to make the burden on plugin developers very light. Under the hood we might want to automatically turn such function annotations into instances of these plugin classes that can then be added to the gui in the same way. These plugins though will ultimately live in different repos.

Also in the long run we might not want any of the `add_plugin` machinery to live on the viewer, which is meant to be more standalone - though we might want to revisit this as we explore the plugins more.

Right now the viewer has been augmented to contain a plugin list that for now is just a normal list, though if multiple plugins get added it will just show the top plugin, and there is no way to swap the active plugin or remove plugins. All this can come, and can follow logic and use machinery similar to the layer list if we think it makes sense.

Right now I have only implemented one plugin - a very simple gaussian blur that takes the image data at layer 0, blurs it and replaces the image data at layer 1. In the future we can allow for named-layers to be selected as inputs and outputs (after #87 merges), and we can allow for a new output layer to be created if non is selected. Similarly we can allow for in place operations with "preview" / "apply" like logic and interactivity. These can be implemented in the base classes (along with the input/output layer selection gui elements) to enable all plugins to use them. Right now there is a very simple "run" button that is provided. The plugin also gets the entire layersList object attached to it when it gets added to the viewer which allows it to select the layers it needs to work with. On the one hand this is quite flexible, on the other maybe it is giving the plugin more than it needs. I don't think the plugin needs access to the whole viewer.

The layout of the plugin gui element is also rather strange and ugly right now, as a box appeneded on the right hand side of the layer list. We'll have to work to make something that looks nice somewhat independently of what gui elements are added in there.

Finally something important to bear is that even quite simple operations can take quite a long time on large images, for example I find the interactivity of the gaussian blur slider already quite laggy on an ~512x512 image. While it is not devastatingly bad, it can be frustrating if I start moving the slider rapidly and everything slows down (particularly if the rest of the gui slows down) suggesting we might want to have separate threads for the gui and the computation and not have the gui get blocked by the computation.

Here's a little gif of the blurring plugin in action:
plugin_blur1

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How has this been tested?

- [x] `examples/layers.ipynb`

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by royerloic** [comment]
That's great Nick. Thanks for this. I think this a great starting point.
I will have some fun in Australia... Let's remember that Napari has its
name because that is the mid--point between Melbourne and San Francisco...
The video looks already cool. But we need indeed to find a better place 
for that plugin panel... ;-)

A few comments:

i) About interactivity: I agree we need to have computation take place in a different thread, maybe check what is the standard approach for thin in PyQT. There might be some machinery to handle 'futures'  and async computation.
ii) One thing to consider is to have the slider only trigger an event when sliding stops, this is something that I have used in Java to avoid triggering computation at every intermediary step.
iii) I am a bit puzzled that gaussian blur is so slow, that sounds very wrong. Do you use conda and its MKL-optimised numpy (that's the default) ? 

Looking forward to play with all this!

Cheers!

Loic

**Comment by jni** [comment]
I presume the slowness comes from computing every intermediate step on that slider:

```python
In [4]: %timeit filters.gaussian(camera, sigma=5)
14.8 ms ± 1.57 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

ie even with a large-ish kernel you are still getting 60fps! =) And that's not even using gputools, coming soon to a skimage near you. ;)

> Under the hood we might want to automatically turn such function annotations into instances of these plugin classes that can then be added to the gui in the same way.

Using Wei's terminology, for functional plugins, I don't want to write a single class, even ourselves as a demo. Seeing `from napari_gui.plugins import GaussianBlur` made me *so* unhappy. Unreasonably so. =P

We do need to figure out the plugin UX. Having a giant panel on any part of the viewer is a non-starter for me. In a sense the viewer is "sacred" so the elements within it shouldn't get resized willy-nilly, only through deliberate actions (e.g. clicking an "expand drawer" button). I think we should have a floating panel, photoshop style, that automatically resizes as the current plugin requires more or fewer parameters. 

As to the plugin experience, here's what my goal is:

- "add plugin": paste a URL, either to a github repo or something privately hosted. Longer term we can add "official" listings of plugins.
- once a plugin has been added, it is forever in the menubar, either under "plugins" or as something more top-level. For this bit we need namespacing worked out. Not to mention a menubar menu. =P ie perhaps as a prerequisite for thinking about plugins we should hash out the menubar first. It's about time Cmd-W worked on Napari windows anyway. =) (We should have a way to give keyboard shortcuts to specific plugins.)
- clicking on the menubar plugin activates the plugin, meaning expanding the floating window or whatever.
- plugins should have a preview tickbox by default.
- what @royerloic said about triggering. We can be more clever and copy the IPython interact idea: sliding triggers Futures, and we keep a dequeue of three of those futures. As new ones get added the old ones get cancelled and deleted, so you're never wasting cycles computing a ton of these.
- layers are very tricky. First, we need a way to assign layers to inputs. This is trivial only in this case, in other cases people will have a shit-ton of layers and they have to select e.g. three as inputs, and then those have to be correctly mapped (in order) to the plugin inputs. So the input panel for every plugin should have the full list of inputs and the layer that is mapped to that input. We can try to be clever but the panel should always be there. Second, I don't know whether it's good to mutate the layers list before actually running the plugin (ie in preview). I suggest that we should have some kind of "virtual layer" functionality so that users can preview without actually touching the layer list. Perhaps the current view would work but (a) it should be a context manager so that the layer list is always preserved even if a plugin crashes, and (b) there should be some indication (e.g. translucence/a blinking/dashed/rotating border) that the layers are not yet final.

I think that about sums up my thoughts. I think this was a valuable exercise because it shows where we need to get to to get a functional plugin, but I don't want to merge this until the plugin is written as I expected, ie:

**Edit:** fixed the below

```
from napari.types import Image, Sigma, Value

def gaussian(image : Image, sigma : Sigma = ..., output : Image = ..., mode : str = ...,
             cval : Value = ..., multichannel : bool = ..., preserve_range : bool = ...,
             truncate : float) -> Image: ...
```

**Comment by jni** [comment]
LOL ignore me for a few mins, Cmd-Enter is "new terminal" on i3 but "submit" on my Mac. 🤦‍♂️

**Comment by jni** [comment]
Ok fixed. =P

**Comment by jni** [comment]
Incidentally, here's the documentation for stub files. tldr we need to mirror the directory structure, maybe even the file structure, of the library that we are wrapping. I think. That'll be super fun for skimage... =P

https://mypy.readthedocs.io/en/latest/stubs.html

**Comment by sofroniewn** [comment]
Thanks @royerloic and @jni for weighing in. Just to reiterate, there are definitely no plans to merge this PR anytime soon (and some of this stuff might not end up living in this repo at all and we might want to go with a different approach that comes in the form of a new PR entirely).

About the blur being slow - if you increase the sigma from 5 to say 50 the blurring becomes proportionally slower - longer than frame rate (now you could say that’s an unrealistically large blur and the slider shouldn’t go so far etc.) but there will often be times when the computation is slow and having the main gui process be blocked seems undesirable (if I add an 100ms sleep things really become painful). I think trying the futures approach that @jni suggested sounds good too, but having two threads might be necessary and could relate to how we think about the app design when accessing remote data etc.

On the UX side of things having a floating panel / panels could work - though I feel that goes in the direction of having popups that @royerloic really doesn’t like. Right now the “plugin panel” is inside a Qslider so you can hide it if you want, but completely agree - making an efficient and visually appealing use of space will be key. If we just standardize layouts and populate things from the top / in standard groups in a fixed panel things could look good - this is how photoshop does their “filters gallery”, which is basically a bunch of plugins - see the right side below

image

As to what it looks like to add plugins and if we’d ever support `from napari_gui.plugins import GaussianBlur` or if we’d only support adding from a URL - I’m really not sure, I think it depends on how we see the major usage modes of Napari, I’ll make an issue about this on Napari-Core where we can have some dedicated discussion there.

As for the only route for plugin creation being function annotation like @jni describes I worry we will be forcing developers down one particular route and we will loose flexibility for creating plugins that have more customized interactivity and complex design - but I’d really have to see what's possible in the other approach to be sure. If you wanted to make a new PR (could be based on this one or not) that implemented the scheme you are suggesting that would be great! 

As to mapping layers - an initial approach could be seeing how many input / output layers of each layer type are required and creating dropdown menus with the relevant layers that can then be selected. @royerloic and I talked about some more complex UI designs using drag / drop lists that we can continue to explore.

Agreed that a preview mode is pretty essential to ultimately allow for “in-place” modifications and that a “virtual-layer” could be a good way to go.

Summing up - I’m going to create a new issue on Napari-Core do discuss usage modes, once the unique layer name PR #87 gets merged I will add layer type specific drop downs to this PR and try and improve the UI, finally maybe @jni and @royerloic when they are together next week will want to work on a new PR that explores the function annotation (non-class) based approach of adding plugins.

**Comment by freeman-lab** [comment]
Great discussion all! Sounds like we'll continue the plugin definition / developer API discussion elsewhere, though at first blush, wearing my plugin developer hat, it seems natural to make both the class-based approach and the functional approach possible and easy :)

On the UX side, super interesting challenge. Here's a mock of my opinionated visual interpretation of @jni's requests, using "drawers" as a compromise between pop ups and floating panels (I don't like multiple independent pop up windows because they feel chaotic, but I agree with @jni  about keeping the viewer sacred -- this tries to strike a balance).

In this flow, you get that GAUSSIAN BLUR drawer to pop up on the right by one of a couple routes, which meet different user stories:
- In the menubar "Add Plugin -> From Source" 
- Click the plus on the right and type in the source
- A call to `add_plugin` or similar in the notebook
- A call in the yet-to-exist console at the bottom

Multiple plugins would occupy multiple little drawers. If you have too many, you can collapse them.

screen shot 2019-02-26 at 1 09 48 pm

File: examples/plugin_class.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
+"""
+Display two image layers using the add_image API
+"""
+
+import sys
+from PyQt5.QtWidgets import QApplication
+
+from skimage import data
+from napari import Window, Viewer
+from napari.plugins import GaussianBlur
```
Comment (jni) [inline examples/plugin_class.py:10]: Here's the vision I want for plugins:

```
from package import module  # contains Napari function annotations
import napari
napari.add_plugin(module)  # autopopulates menus with module functions
```

This works for functional modules, and maybe also for GUI modules. Without this, there is virtually zero chance of getting to the goal of napari recording pure Python. It's feasible, but it will never be Python that anyone would want to write.

Comment (sofroniewn) [reply examples/plugin_class.py:10]: ah ok, this is nice to see - helps me understand. I definitely think we want to support this for the reasons you mention, and see how for large libraries with lots of functions that you want a menu bar to populate it makes sense. I'm just a little unsure how we will make the corresponding GUI elements for when those functions require additional parameters, but maybe this is doable and we just need to keep exploring.

If this becomes the main way to get custom code into napari then I might suggest that we "re-brand" our "plugin ecosystem" to a python "module ecosystem" and talk about `napari.add_module`

Comment (jni) [reply examples/plugin_class.py:10]: > I'm just a little unsure how we will make the corresponding GUI elements for when those functions require additional parameters

The idea is to automatically determine the GUI elements from the input types. We had a discussion about this where @freeman-lab pointed us to such a framework for Javascript. 

> we "re-brand" our "plugin ecosystem" to a python "module ecosystem" and talk about `napari.add_module`

yes, that's exactly the right idea imho! Napari headless is just Python! =)

**Comment by sofroniewn** [comment]
I'm going to close this PR now as the code base and our thinking about plugins has moved further along. The vision from @jni of plugins used as follows resonates

```python
from package import module  # contains Napari function annotations
import napari
napari.add_plugin(module)  # autopopulates menus with module functions
```

As is being discussed in #229 this repo will not contain any plugin related code. Instead - the `napari-app` repo will contain the plugin code, and this discussion can continue there when the time is right
